### PR TITLE
ARROW-4036: [C++] Pluggable Status message, by exposing an abstract delegate class.

### PR DESCRIPTION
**Pull request by emkornfield**
This provides less "pluggability" but I think still offers a clean model for extension (subsystems can wrap the constructor for there purposes, and provide external static methods to check for particular types of errors).

### CODE REVIEW

**Comment by emkornfield** [comment]
This also assumes subsystem codes don't map to top level codes (I think I would prefer they do, but I could see them going both ways).

File: cpp/src/arrow/status.h
Comment (pitrou) [inline cpp/src/arrow/status.h]: I don't think that's the right approach. If I have a Python `ValueError`, I want it both to map to `StatusCode::Invalid` _and_ to retain the original Python exception.

Comment (emkornfield) [reply cpp/src/arrow/status.h]: Yep, I tend to agree.  So two issues to improve on:
*  top-level codes should be orthogonal from subsytem codes
+  Potentially need the ability to directly link to underlying exception (e.g. PyObject for a python error)

Any other high level concepts/functionallty missing?

Comment (pitrou) [reply cpp/src/arrow/status.h]: Right. I also think subsystem codes can be optional (for example, Python would just use the underlying exception instead of defining integer codes).

So instead of having a subsystem code, we could have an opaque optional detail instance, e.g.:
```cpp
class StatusDetail {
 public:
  virtual ~StatusDetail();
  virtual std::string ToString() const = 0;
};
```

File: cpp/src/arrow/status.h
Comment (lidavidm) [inline cpp/src/arrow/status.h]: Does Flight also warrant a subsystem? It would be nice to express errors like unauthorized, service unavailable/overloaded, etc. (tagging my work account @lihalite)

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: Yeah, I came to the same conclusion after stepping away from the computer. Will try to update tonight.

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: Going to remove this abd go with the approach suggested by Antoine

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: I've updated to just have a status detail, that wraps has a message accessor and unify the current implementation with a private implementation of the interface.  I can see arguments for keeping a separate message member as well, so I'm happy to add it back.

**Comment by emkornfield** [comment]
If you want me to try to cleanup one of the subsystem (python?) in this PR let me know and I can give it a shot.

File: cpp/src/arrow/status.h
Comment (pitrou) [reply cpp/src/arrow/status.h]: I would be in favour of having a separate message member, so that the detail can remain null if there's no error-specific detail to carry.

Comment (emkornfield) [reply cpp/src/arrow/status.h]: Added msg field. back.

**Review by pitrou — COMMENTED** [review]
I think this looks nice on the principle. It would be useful to validate this by migrating the Gandiva, Plasma... status codes.

File: cpp/src/arrow/status-test.cc
Comment (emkornfield) [reply cpp/src/arrow/status-test.cc]: added "." not sure how many detail messages will add it themselves.

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: Added with_detail.

**Comment by emkornfield** [comment]
> I think this looks nice on the principle. It would be useful to validate this by migrating the Gandiva, Plasma... status codes.

Will give this a try with plasma

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: yes.

**Comment by emkornfield** [comment]
@pitrou this demonstrates using the the StatusDetail with Python and Plasma. This breaks some C++ and Python APIs, I'm not sure if we are OK with that.

If this approach is OK, I will do a follow-up PR to remove Gandiva codes as well.

CC @pcmoritz

**Comment by emkornfield** [comment]
putting back in WIP status until I can get CI green (python tests passed for me locally)

**Comment by emkornfield** [comment]
@pitrou after a lot of debugging I figured out why the builds were failing only to discover you had fixed with with your PR to provide better errors (there was a copy of Status in a macro that I did not find initally).  Thank you!  I think this is ready to review.  I'll file a follow-up JIRA once this is merged to migrade Gandiva away from this as well.

**Comment by pitrou** [comment]
Will review, thank you :-)

**Comment by emkornfield** [comment]
I'm looking into the lint failure, but I'm guessing i wll need to make some changes anyways.

**Comment by pitrou** [comment]
Yes, don't worry about the lint failure for now.

File: cpp/src/arrow/status.cc
Comment (emkornfield) [inline cpp/src/arrow/status.cc]: bad merge.  will remove this.

File: cpp/src/arrow/compute/kernels/cast.cc
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/cast.cc]: Wow, I didn't know `cast.cc` redefined its own error-handling macros...

**Comment by pitrou** [comment]
@emkornfield I think I'll update the Python parts directly in your branch.

File: cpp/src/arrow/python/common.cc
Comment (pitrou) [inline cpp/src/arrow/python/common.cc]: Do you think we should use `strcmp` instead? Or is it on purpose?

File: cpp/src/arrow/compute/kernels/cast.cc
Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/cast.cc]: yeah.  Somehow I missed the same issue in Status, which took a lot of time to find :/

File: cpp/src/arrow/python/common.cc
Comment (emkornfield) [reply cpp/src/arrow/python/common.cc]: This is on purpose as it should be cheap and I don't think we want to allow "impersonation".  it might be better to type-id be a void* or I can go back to using RTTI (check for nullptr after casting to expected type).   Thoughts?

**Comment by pitrou** [comment]
@emkornfield I'm taking over the Python parts.

**Comment by emkornfield** [comment]
> @emkornfield I think I'll update the Python parts directly in your branch.

@pitrou sounds good, hopefully it isn't too ugly :(  I could have swarn I recompiled after I made the lint fixes, sorry my attention to detail has been lacking in general lately.

**Comment by emkornfield** [comment]
@pitrou OK, I'll stop pusing

File: cpp/src/arrow/status.h
Comment (emkornfield) [reply cpp/src/arrow/status.h]: @lihalite when this lands, I think some codes should probably be top level (unavailable seems generic across multiple service type things) but some might be flight specific.

Comment (lihalite) [reply cpp/src/arrow/status.h]: Ah ok, great. I started sketching out a set of codes [for Flight/Java](https://github.com/lihalite/arrow/pull/7/files#diff-dbd9be8475cffe0a29ae8ce317320cc0), I can pick up the C++/Python side too once this lands.

File: cpp/src/arrow/python/common.cc
Comment (pitrou) [reply cpp/src/arrow/python/common.cc]: Pointer comparison is fine with me. However, I think we should also use more standardized names, e.g. the fully-qualified class name as I've done in Python (`arrow::py::PythonErrorDetail`).

I don't know if we support building Arrow without RTTI. There are some `dynamic_cast`s in the codebase. See especially `InheritsFrom` in `src/arrow/compute/expression.h`.

**Comment by pitrou** [comment]
I've done my thing on Python. Let's wait for CI a bit.

**Comment by pitrou** [comment]
@ursabot build test

**Comment by pitrou** [comment]
@ursabot build

**Comment by kszucs** [comment]
@ursabot build

**Comment by pitrou** [comment]
The failed ursabot builds are unrelated, @kszucs is busy doing thins on BuildBot.
The failed AppVeyor build is a transient download failure

Travis-CI build: https://travis-ci.org/pitrou/arrow/builds/553376058
AppVeyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/25698396

File: cpp/src/arrow/compute/kernels/cast.cc
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/cast.cc]: Perhaps open a JIRA to factor those out?

File: cpp/src/plasma/common.h
Comment (pitrou) [inline cpp/src/plasma/common.h]: Would probably be less confusing to follow the `Status(code, message)` convention rather than reverting it.

File: cpp/src/plasma/common.cc
Comment (pitrou) [inline cpp/src/plasma/common.cc]: Turn this into `plasma::PlasmaStatusDetail`?

File: python/pyarrow/__init__.py
Comment (pitrou) [inline python/pyarrow/__init__.py]: We shouldn't import plasma from here. People will have to import those exceptions from `arrow.plasma` instead.

**Review by pitrou — CHANGES_REQUESTED** [review]
A couple more things to change. I can tackle those tomorrow if you like.

File: cpp/src/arrow/compute/kernels/cast.cc
Comment (emkornfield) [reply cpp/src/arrow/compute/kernels/cast.cc]: https://issues.apache.org/jira/browse/ARROW-5833

File: cpp/src/arrow/python/common.cc
Comment (emkornfield) [inline cpp/src/arrow/python/common.cc]: thanks for doing this, I started down this path and decided to keep it simpler because there were semantics of the GIL I was hazy on.

File: cpp/src/plasma/common.cc
Comment (emkornfield) [reply cpp/src/plasma/common.cc]: Yep

File: cpp/src/plasma/common.h
Comment (emkornfield) [reply cpp/src/plasma/common.h]: Fixing it.

File: python/pyarrow/__init__.py
Comment (emkornfield) [reply python/pyarrow/__init__.py]: OK, I was trying to maintain a little backwards compatibility but i'll move them.

Comment (emkornfield) [reply python/pyarrow/__init__.py]: removed.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
I'm going to merge once CI is green.

**Pull request closed without merging**
